src/tree/mod.rs

```diff
@@ -289,105 +246,162 @@ impl RatchetTree {
+    /// > intermediate nodes in the path above the leaf. The path is ordered
+    /// > from the closest node to the leaf to the root; each node MUST be the
+    /// > parent of its predecessor.
+    ///
+    pub(crate) fn update_path(
         &mut self,
         sender: LeafIndex,
-        direct_path: &DirectPath,
+        update_path: &UpdatePath,
         group_context: &[u8],
-    ) -> CommitSecret {
+    ) -> Result<CommitSecret, TreeError> {
         let own_index = self.get_own_node_index();
-        // TODO check that the direct path is long enough
 
         // Find common ancestor of own leaf and sender leaf
-        let common_ancestor = treemath::common_ancestor(NodeIndex::from(sender), own_index);
+        let common_ancestor_index =
+            treemath::common_ancestor_index(NodeIndex::from(sender), own_index);
 
-        // Calculate sender direct path & copath, common path
-        let sender_dirpath = treemath::dirpath_root(NodeIndex::from(sender), self.leaf_count());
-        let sender_copath = treemath::copath(NodeIndex::from(sender), self.leaf_count());
+        // Calculate sender direct path & co-path, common path
+        let sender_direct_path =
+            treemath::direct_path_root(NodeIndex::from(sender), self.leaf_count());
+        let sender_co_path = treemath::copath(NodeIndex::from(sender), self.leaf_count());
 
         // Find the position of the common ancestor in the sender's direct path
-        let common_ancestor_sender_dirpath_index = sender_dirpath
+        let common_ancestor_sender_dirpath_index = sender_direct_path
             .iter()
-            .position(|x| *x == common_ancestor)
+            .position(|&x| x == common_ancestor_index)
             .unwrap();
-        let common_ancestor_copath_index = sender_copath[common_ancestor_sender_dirpath_index];
+        let common_ancestor_copath_index =
+            match sender_co_path.get(common_ancestor_sender_dirpath_index) {
+                Some(i) => i.clone(),
+                None => return Err(TreeError::InvalidArguments),
+            };
 
-        // Resolve the node of that copath index
+        // Resolve the node of that co-path index
         let resolution = self.resolve(common_ancestor_copath_index);
-        let position_in_resolution = resolution.iter().position(|x| *x == own_index).unwrap_or(0);
-        // TODO Check resolution.len() == encrypted_path_secret.len()
+        let position_in_resolution = resolution.iter().position(|&x| x == own_index).unwrap_or(0);
 
         // Decrypt the ciphertext of that node
-        let hpke_ciphertext = &direct_path.nodes[common_ancestor_sender_dirpath_index]
-            .encrypted_path_secret[position_in_resolution];
+        let common_ancestor_node = match update_path.nodes.get(common_ancestor_sender_dirpath_index)
+        {
+            Some(node) => node,
+            None => return Err(TreeError::InvalidArguments),
+        };
+        assert_eq!(
```

Comment:
Shouldn't this be a `debug_assert()`? Otherwise the following code wouldn't get executed.

Reply:
It should!

Reply:
There a few more places where that's the case, should I flag them?

Reply:
I replace all in here.